libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/MapboxRouteLineApi.kt

```diff
@@ -730,34 +760,166 @@ class MapboxRouteLineApi(
-                // need to be rendered.
-                null -> highlightActiveLeg(routeProgress, consumer)
-                else -> {
-                    ifNonNull(routeProgress.currentLegProgress) { routeLegProgress ->
-                        if (routeLegProgress.legIndex > activeLegIndex) {
-                            jobControl.scope.launch(Dispatchers.Main) {
-                                mutex.withLock {
-                                    alternativelyStyleSegmentsNotInLeg(
-                                        routeLegProgress.legIndex,
-                                        routeLineExpressionData
-                                    )
-                                    activeLegIndex = routeLegProgress.legIndex
+        when {
+            routeLineOptions.styleInactiveRouteLegsIndependently -> {
+                when (routeLineOptions.vanishingRouteLine) {
+                    // If the styleInactiveRouteLegsIndependently feature is enabled but the
+                    // vanishingRouteLine feature is not enabled then side effects are generated and
+                    // need to be rendered.
+                    null -> highlightActiveLeg(
+                        routeProgress,
+                        routeLineMaskingLayerDynamicData,
+                        consumer
+                    )
+                    else -> {
+                        ifNonNull(routeProgress.currentLegProgress) { routeLegProgress ->
+                            if (routeLegProgress.legIndex > activeLegIndex) {
+                                jobControl.scope.launch(Dispatchers.Main) {
+                                    mutex.withLock {
+                                        alternativelyStyleSegmentsNotInLeg(
+                                            routeLegProgress.legIndex,
+                                            routeLineExpressionData,
+                                            routeLineOptions
+                                                .resourceProvider
+                                                .routeLineColorResources
+                                                .inActiveRouteLegsColor
+                                        )
+                                    }
                                 }
                             }
                         }
+                        provideRouteLegUpdate(routeLineMaskingLayerDynamicData, consumer)
                     }
                 }
             }
+            else -> provideRouteLegUpdate(routeLineMaskingLayerDynamicData, consumer)
+        }
+    }
+
+    private fun provideRouteLegUpdate(
+        routeLineOverlayDynamicData: RouteLineDynamicData?,
+        consumer: MapboxNavigationConsumer<Expected<RouteLineError, RouteLineUpdateValue>>
+    ) {
+        ifNonNull(routeLineOverlayDynamicData) { maskingLayerData ->
+            getRouteDrawData { expected ->
```

Comment:
How expensive is this function call for very long routes? We ultimately only need the dynamic data from it, we don't need the geometries and all the rest. The vanishing update functions are able to get the dynamic data without rebuilding geometries so maybe we could try doing the same here, unless it's cheap to get all of it of course.

Reply:
With all the caching in the class calling this shouldn't be that expensive. It's only called when there is a transition from one route leg to another.

I'm going from memory here but I think the main reason I had to do this was because the data type returned (RoutelineUpdateValue) has non-nullable primary and alternative `RouteLineDyamicData` fields.  I had to get data from somewhere and `getRouteDrawData` seemed like an obvious source.  Because of `RouteLineUpdateValue` we're painted into a bit of a corner here.